raft/log.go

```diff
@@ -1021,9 +1026,11 @@ func (l *Log) candidateLoop(closing <-chan struct{}) State {
 			l.lock()
 			l.mustSetTermIfHigher(hb.term)
 			if hb.term >= l.term {
+				l.tracef("candidateLoop: new leader: old=%d new=%d", l.leaderID, hb.leaderID)
 				l.leaderID = hb.leaderID
 			}
 			l.unlock()
+			return Follower
```

Comment:
Are you certain? This is critical.

A node may *think* it's a leader and send out a heartbeat. If this was not the case, there should be no need for the code `if hb.term >= l.term`. To quote the paper: "If the term in the RPC is smaller than the candidate’s current term, then the candidate rejects the RPC and continues in candidate state." If what you say is correct, then putting the return inside `if` block will make no difference because if this code is only receiving heartbeats from the "real" leader then the term will always be equal to or greater than the current term.

So put it inside the `if` block, it should make no difference.